doc/design/service-discovery.md

```diff
@@ -0,0 +1,316 @@
+
+### Decision
+
+The **event-based notification alternative D** fulfills both requirements for a asychronous, one shot-polling approach as
+well as the event-based one. It can replace the `InterfacePort`s, which broadens the usage e.g. for gateways or even
+introspection 2.0. Furthermore, alternative D does not add any new non-trivial building blocks. The one-shot RPC call
+does not lead to a shared memory transmission compared to the alternative B (request/response feature usage) as the
+complete service registry is stored locally.
+
+### Code example
+
+#### Event-based notification: Alternative B
+
+```cpp
+auto& runtime = PoshRuntime::initRuntime("myApp");
+auto* interfacePortData = runtime.getMiddlewareInterface(capro::Interfaces::INTERNAL);
+
+InterfacePort caproSubscriber(interfacePortData);
+Listener myListner;
+
+void onDiscoveryUpdateCallback(InterfacePort* subscriber)
+{
+    subscriber->tryGetCaProMessage().and_then([](auto& caproMessage){
+        // Has any service that is relevant for me changed?
+        if(caproMessage.m_serviceDescription == myRelevantServiceDescription && someOtherCondition())
+        {
+            // call user-defined callback
+        }
+    });
+}
+
+myListner.attachEvent(caproSubscriber, DATA_RECEIVED, createNotificationCallback(onDiscoveryUpdateCallback));
+```
+
+#### Event-based notification: Alternative D
+
+```cpp
+class DiscoveryInfo
+{
+  public:
+
+    // Move all service-related methods from PoshRuntime to this class
+    cxx::expected<InstanceContainer, FindServiceError>
+    PoshRuntime::findService(const capro::ServiceDescription& serviceDescription) noexcept
+    {
+        // Update local service registry
+        get();
+
+        return filterFor(serviceDescription);
+    }
+
+    // @todo check if still needed
+    bool offerService(const capro::ServiceDescription& serviceDescription) noexcept;
+    void stopOfferService(const capro::ServiceDescription& serviceDescription) noexcept;
+
+    ServiceRegistryTopic get() noexcept
+    {
+        m_subscriber.take().and_then([](auto& ServiceRegistryTopic){
+            // Update our local copy of the service registry
+            m_lastServiceRegistry = ServiceRegistryTopic;
+        }).or_else([](){
+            // No change in service registry
+        });
+        return m_lastServiceRegistry;
+    }
+
+    ServiceRegistryTopic getDelta() noexcept
+    {
+        auto old = m_lastServiceRegistry;
+        get();
+        return old - m_lastServiceRegistry;
+    }
+
+    ServiceRegistryTopic getDelta(const capro::ServiceDescription& serviceDescription) noexcept;
+
+   filterFor(const capro::ServiceDescription& serviceDescription) noexcept;
+  private:
+    Subscriber<ServiceRegistryTopic> m_subscriber{"ServiceDiscovery", "GlobalInstance", "ServiceRegistryTopic"};
+    ServiceRegistryTopic m_lastServiceRegistry;
+}
+
+// Class should be attachable to listner
```

Comment:
With these combined the user can create an active class which handles the state update automatically when new data is received. The callback must contain the required state update on receiving the new registry.

While the user can do so on its own, it may be useful to define a `DynamicDiscovery` (or similar) class which encapsulates it for he default use case. If the user wants do perfom somemthing specific he needs to either write its own version or  provide a different callback.

Reply:
> While the user can do so on its own, it may be useful to define a DynamicDiscovery (or similar) class which encapsulates it for he default use case.
Will think about it, although I have the feeling this should probably be done in the binding code (e.g. AUTOSAR binding or rmw)